### PR TITLE
feat: add tree-sitter-ng for tree generator.

### PR DESCRIPTION
**Pull request by XYZboom**
tree-sitter-ng is a java binding of tree-sitter, so with tree-sitter-ng, we don't need python anymore and we can generate tree in memory instead of reading xml from disk.

### CODE REVIEW

**Comment by XYZboom** [comment]
fix a bug and add some unit tests for this fix.

**Comment by jrfaller** [comment]
Hi! Thanks a lot for the bugfix. Did you get the chance to look at my comments? Cheers!

**Comment by XYZboom** [comment]
Sorry, I can't seem to see your comments. Can you help me find it, thanks!

**Comment by XYZboom** [comment]
I have already checked the option to open comments, but the interface hasn't changed.
image

File: gen.treesitter-ng/build.gradle [hunk 1]
```diff
@@ -0,0 +1,22 @@
+description = 'GumTree tree generator for tree-sitter-ng.'
+
+ext.isNative = true
+
+dependencies {
+    implementation("org.yaml:snakeyaml:2.0")
+    implementation("io.github.bonede:tree-sitter-c:0.20.5")
+    implementation("io.github.bonede:tree-sitter-cmake:0.4.1")
+    implementation("io.github.bonede:tree-sitter-c-sharp:0.20.0")
+    implementation("io.github.bonede:tree-sitter:0.22.2")
```
Comment (jrfaller) [inline gen.treesitter-ng/build.gradle:10]: Maybe this dependency should go just below snakeyaml?

**Review by jrfaller — CHANGES_REQUESTED** [review]
There are a few minor issues, and a bigger one concerning the selector matching.

**Comment by jrfaller** [comment]
Sorry @XYZboom  I was thinking you were seeing them 😢 . I assigned you to the PR and clicked on a menu to finish the review to ask for changes. Do you see them now? Cheers!

**Comment by XYZboom** [comment]
> Sorry @XYZboom I was thinking you were seeing them 😢 . I assigned you to the PR and clicked on a menu to finish the review to ask for changes. Do you see them now? Cheers!

Thank you, I can now see these comments!

File: gen.treesitter-ng/src/main/java/com/github/gumtreediff/gen/treesitterng/AbstractTreeSitterNgGenerator.java [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+    @SuppressWarnings("unchecked")
+    protected static Pair<Tree, Boolean> tsNode2GumTree(
+            List<String> contentLines, Map<String, Object> currentRule, TreeContext context, TSNode node) {
+        String type = node.getType();
+        TSNode parent = node.getParent();
+        String parentAndChildType = parent.isNull() ? type : (parent.getType() + " " + type);
+        String label = getLabel(contentLines, node);
+        if (currentRule.containsKey(YAML_IGNORED)) {
+            List<String> ignores = (List<String>) currentRule.get(YAML_IGNORED);
```
[the remark this answers is not in the document]
Comment (XYZboom) [reply gen.treesitter-ng/src/main/java/com/github/gumtreediff/gen/treesitterng/AbstractTreeSitterNgGenerator.java:159]: Add a method `matchNodeOrAncestorTypes` to handle arbitrary long chain of parents.

File: gen.treesitter-ng/build.gradle (hunk 1 above)
Comment (XYZboom) [reply gen.treesitter-ng/build.gradle]: Yes, it should. It's the core module of tree-sitter-ng.

File: gen.treesitter-ng/src/main/java/com/github/gumtreediff/gen/treesitterng/AbstractTreeSitterNgGenerator.java [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+            tree = context.createTree(TypeSet.type(type), label);
+        } else {
+            tree = context.createTree(TypeSet.type(type));
+        }
+        tree.setPos(calculateOffset(contentLines, node.getStartPoint()));
+        tree.setLength(label.length());
+        return new Pair<>(tree, flatten);
+    }
+
+    private static TreeContext generateFromTressSitterTree(
```
[the remark this answers is not in the document]
Comment (XYZboom) [reply gen.treesitter-ng/src/main/java/com/github/gumtreediff/gen/treesitterng/AbstractTreeSitterNgGenerator.java]: Fix in new commits, thanks!

File: benchmark/datasets
[the remark this answers is not in the document]
Comment (XYZboom) [reply benchmark/datasets:1]: I'm not sure why this change appeared in this PR, maybe it's a bug on GitHub. When I clicked on this change, there wasn't any commits I made to this change of submodule😢. @jrfaller 
image
image
In the 2nd screenshot above, none of this commits is made of me, perhaps it'a a bug?

File: gen.treesitter/src/main/java/com/github/gumtreediff/gen/treesitter/AbstractTreeSitterGenerator.java
[the remark this answers is not in the document]
Comment (XYZboom) [reply gen.treesitter/src/main/java/com/github/gumtreediff/gen/treesitter/AbstractTreeSitterGenerator.java:1]: Sorry, I rolled it back

**Pull request merged**